apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/StorageClientImpl.DeleteBucket.cs

```diff
@@ -23,37 +24,79 @@ public sealed partial class StorageClientImpl : StorageClient
     {
         /// <inheritdoc />
         public override void DeleteBucket(string bucket, DeleteBucketOptions options = null) =>
-            CreateDeleteBucketRequest(bucket, options).Execute();
+            DeleteBucketImpl(ValidateBucketName(bucket), options);
 
         /// <inheritdoc />
         public override Task DeleteBucketAsync(
             string bucket,
             DeleteBucketOptions options = null,
-            CancellationToken cancellationToken = default
-        ) => CreateDeleteBucketRequest(bucket, options).ExecuteAsync(cancellationToken);
+            CancellationToken cancellationToken = default) =>
+            DeleteBucketAsyncImpl(ValidateBucketName(bucket), options, cancellationToken);
 
         /// <inheritdoc />
-        public override void DeleteBucket(Bucket bucket, DeleteBucketOptions options = null)
-            => CreateDeleteBucketRequest(bucket, options).Execute();
+        public override void DeleteBucket(Bucket bucket, DeleteBucketOptions options = null) =>
+            DeleteBucketImpl(ValidateBucket(bucket, nameof(bucket)), options);
 
         /// <inheritdoc />
         public override Task DeleteBucketAsync(
             Bucket bucket,
             DeleteBucketOptions options = null,
-            CancellationToken cancellationToken = default
-        ) => CreateDeleteBucketRequest(bucket, options).ExecuteAsync(cancellationToken);
+            CancellationToken cancellationToken = default) =>
+            DeleteBucketAsyncImpl(ValidateBucket(bucket, nameof(bucket)), options, cancellationToken);
 
-        private BucketsResource.DeleteRequest CreateDeleteBucketRequest(Bucket bucket, DeleteBucketOptions options)
+        // Implementation methods assuming a previously-validate bucket name
+
+        private void DeleteBucketImpl(string bucket, DeleteBucketOptions options)
         {
-            ValidateBucket(bucket, nameof(bucket));
-            var request = Service.Buckets.Delete(bucket.Name);
-            options?.ModifyRequest(request);
-            return request;
+            // Create the request at the start, so we validate the options before we
+            // start deleting objects.
+            var deleteBucketRequest = CreateDeleteBucketRequest(bucket, options);
+            string userProject = options?.UserProject;
+            if (options?.DeleteObjects ?? false)
+            {
+                // Don't start deleting objects if the bucket has the wrong metageneration to start with.
+                var preconditionOptions = options?.CreateGetBucketOptionsForPreconditions();
```

Comment:
Don't we know `options` is not null here?

Reply:
Yes, we do - doh.

Reply:
(Now fixed.)